# 시각화

In [2]:
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set() # setting seaborn default for plots

In [3]:
import pandas as pd

train=pd.read_csv('./train.csv')
test=pd.read_csv('./test_x.csv')

# 데이터 전처리

In [4]:
import numpy as np
family_drop_score=10

train['familysize']=np.where(train['familysize'] >= family_drop_score, family_drop_score, train['familysize'])
test['familysize']=np.where(test['familysize'] >= family_drop_score, family_drop_score, test['familysize'])

train['education']=np.where(train['education'] == 0, 2, train['education'])
test['education']=np.where(test['education'] == 0, 2, test['education'])

train['urban']=train['urban'].astype('str')
test['urban']=test['urban'].astype('str')

train['education']=train['education'].astype('str')
test['education']=test['education'].astype('str')

# 원핫 인코딩 (gender, race, religion)

In [5]:
train=pd.get_dummies(train, columns=['race'])
train=pd.get_dummies(train, columns=['religion'])
train=pd.get_dummies(train, columns=['urban'])
train=pd.get_dummies(train, columns=['hand'])
train=pd.get_dummies(train, columns=['gender'])

test=pd.get_dummies(test, columns=['race'])
test=pd.get_dummies(test, columns=['religion'])
test=pd.get_dummies(test, columns=['urban'])
test=pd.get_dummies(test, columns=['hand'])
test=pd.get_dummies(test, columns=['gender'])

# 라벨 인코딩 (gender, race, religion)

In [6]:
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()
train['age_group_code'] = label.fit_transform(train['age_group'])
train['education_code'] = label.fit_transform(train['education'])

test['age_group_code'] = label.fit_transform(test['age_group'])
test['education_code'] = label.fit_transform(test['education'])

# Feature : Mach_score (마키아벨리니즘 스코어)

In [7]:
Answers = ['QaA', 'QbA', 'QcA', 'QdA', 'QeA',
             'QfA', 'QgA', 'QhA', 'QiA', 'QjA', 
             'QkA', 'QlA', 'QmA', 'QnA', 'QoA', 
             'QpA', 'QqA', 'QrA', 'QsA', 'QtA']

In [8]:
reverse = ["QeA", "QfA", "QkA", "QqA", "QrA"]
for r in reverse: 
        train[r] = 6 - train[r]
        test[r] = 6 - test[r]

In [9]:
reverse = ["QaA", "QdA", "QgA", "QiA", "QnA"]
for r in reverse: 
        train[r] = 6 - train[r]
        test[r] = 6 - test[r]

In [10]:
train['Mach_score'] = train[Answers].mean(axis = 1)
test['Mach_score'] = test[Answers].mean(axis = 1)

# TIPI
## 성실성 (Conscientiousness)
## 우호성 (Agreeableness)
## 신경성 (Neuroticism) - 정서적 안정감
## 개방성 (Openness to experience)
## 외향성 (Extroversion)

- 성실성 : {3번 점수 + (8 - '8번 점수')} ÷ 2
- 우호성 : {7번 점수 + (8 - '2번 점수')} ÷ 2
- 정서적 안정성(점수가 낮으면 신경성과 관련) : {9번 점수 + (8 - '4번 점수')} ÷ 2
- 개방성 : {5번 점수 + (8 - '10번 점수')} ÷ 2
- 외향성 : {1번 점수 + (8 - '6번 점수')} ÷ 2

In [11]:
train['TIPT_C']=(train['tp03']+(8-train['tp08']))/2
train['TIPT_A']=(train['tp07']+(8-train['tp02']))/2
train['TIPT_N']=(train['tp09']+(8-train['tp04']))/2
train['TIPT_O']=(train['tp05']+(8-train['tp10']))/2
train['TIPT_E']=(train['tp01']+(8-train['tp06']))/2

test['TIPT_C']=(test['tp03']+(8-test['tp08']))/2
test['TIPT_A']=(test['tp07']+(8-test['tp02']))/2
test['TIPT_N']=(test['tp09']+(8-test['tp04']))/2
test['TIPT_O']=(test['tp05']+(8-test['tp10']))/2
test['TIPT_E']=(test['tp01']+(8-test['tp06']))/2

# wf_(1~3) : 설문자의 어휘 능력 (허구인 단어의 정의를 앎)

In [12]:
wf=['wf_01', 'wf_02', 'wf_03']

train['wf']=train[wf].mean(axis=1)
test['wf']=test[wf].mean(axis=1)

# drop 할 feature들

In [13]:
drop_feature=[
#                 'engnat',
    
#                 'familysize',
    
#                 'married',
    
#                 'race_Arab', 'race_Asian','race_Black','race_Indigenous Australian','race_Native American','race_Other','race_White',               
    
#                 'religion_Agnostic','religion_Atheist', 'religion_Buddhist','religion_Christian_Catholic','religion_Christian_Mormon', 
#                 'religion_Christian_Other', 'religion_Christian_Protestant', 'religion_Hindu','religion_Jewish','religion_Muslim', 
#                 'religion_Other','religion_Sikh',
    
#                 'gender_Female','gender_Male',
                
#                 'urban_0','urban_1','urban_2','urban_3',
    
                'education',
                
                'age_group',
                
                'QaA', 'QbA', 'QcA', 'QdA', 'QeA','QfA', 'QgA', 'QhA', 'QiA', 'QjA', 'QkA', 'QlA', 'QmA', 'QnA', 'QoA','QpA', 
                'QqA', 'QrA', 'QsA', 'QtA',
                
                'QaE', 'QbE', 'QcE', 'QdE', 'QeE', 'QfE', 'QgE', 'QhE', 'QiE', 'QjE', 'QkE', 'QlE','QmE', 'QnE', 'QoE', 'QpE', 
                'QqE', 'QrE', 'QsE', 'QtE',
              
                'tp01', 'tp02', 'tp03', 'tp04', 'tp05','tp06','tp07', 'tp08', 'tp09', 'tp10',
              
#                 'wr_01', 'wr_02', 'wr_03', 'wr_04', 'wr_05', 'wr_06', 'wr_07', 'wr_08', 'wr_09','wr_10','wr_11', 'wr_12', 'wr_13', 
                'wr_02',  'wr_04', 'wr_07', 'wr_08', 'wr_10', 'wr_12', 'wr_13', 
                'wf_01', 'wf_02', 'wf_03', 'wf',
              
                'hand_0','hand_1','hand_2','hand_3',
              
              ]

train = train.drop(drop_feature, axis = 1)
test = test.drop(drop_feature, axis = 1)

# Model 생성

In [14]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
import lightgbm as lgb

# 사용한 feature 목록

In [15]:
for i in train:
    print(i)


index
engnat
familysize
married
voted
wr_01
wr_03
wr_05
wr_06
wr_09
wr_11
race_Arab
race_Asian
race_Black
race_Indigenous Australian
race_Native American
race_Other
race_White
religion_Agnostic
religion_Atheist
religion_Buddhist
religion_Christian_Catholic
religion_Christian_Mormon
religion_Christian_Other
religion_Christian_Protestant
religion_Hindu
religion_Jewish
religion_Muslim
religion_Other
religion_Sikh
urban_0
urban_1
urban_2
urban_3
gender_Female
gender_Male
age_group_code
education_code
Mach_score
TIPT_C
TIPT_A
TIPT_N
TIPT_O
TIPT_E


# 모델 전처리

In [16]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import *
from lightgbm import LGBMClassifier

import warnings
import gc
warnings.filterwarnings("ignore")

X = train.copy()
X.drop('voted', axis=1, inplace = True)
Y = train['voted']

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

In [17]:
import lightgbm as lgb

model=LGBMClassifier(
n_estimators=1000,
learning_rate=0.01,
boosting='rf',
objective= 'binary',
metric= 'binary_logloss',
bagging_freq =1,
bagging_fraction = 0.5,
is_training_metric= True,
seed=2020
)
evals=[(x_test, y_test)]
model.fit(x_train, y_train, early_stopping_rounds=100, eval_metric="AUC",eval_set=evals, verbose=True)

[LightGBM] [Warning] boosting is set=rf, boosting_type=gbdt will be ignored. Current value: boosting=rf
[LightGBM] [Warning] bagging_fraction is set=0.5, subsample=1.0 will be ignored. Current value: bagging_fraction=0.5
[LightGBM] [Warning] bagging_freq is set=1, subsample_freq=0 will be ignored. Current value: bagging_freq=1
[1]	valid_0's auc: 0.751284	valid_0's binary_logloss: 0.575854
[2]	valid_0's auc: 0.75558	valid_0's binary_logloss: 0.573841
[3]	valid_0's auc: 0.757808	valid_0's binary_logloss: 0.572958
[4]	valid_0's auc: 0.760064	valid_0's binary_logloss: 0.571674
[5]	valid_0's auc: 0.761428	valid_0's binary_logloss: 0.570643
[6]	valid_0's auc: 0.762164	valid_0's binary_logloss: 0.570362
[7]	valid_0's auc: 0.762851	valid_0's binary_logloss: 0.569833
[8]	valid_0's auc: 0.763161	valid_0's binary_logloss: 0.569616
[9]	valid_0's auc: 0.762871	valid_0's binary_logloss: 0.569717
[10]	valid_0's auc: 0.763015	valid_0's binary_logloss: 0.569612
[11]	valid_0's auc: 0.763051	valid_0's bi

[159]	valid_0's auc: 0.764196	valid_0's binary_logloss: 0.569096
[160]	valid_0's auc: 0.764179	valid_0's binary_logloss: 0.5691
[161]	valid_0's auc: 0.764181	valid_0's binary_logloss: 0.569104
[162]	valid_0's auc: 0.764188	valid_0's binary_logloss: 0.569097
[163]	valid_0's auc: 0.764194	valid_0's binary_logloss: 0.569096
[164]	valid_0's auc: 0.764209	valid_0's binary_logloss: 0.569091
[165]	valid_0's auc: 0.764216	valid_0's binary_logloss: 0.569087
[166]	valid_0's auc: 0.764225	valid_0's binary_logloss: 0.569081
[167]	valid_0's auc: 0.764236	valid_0's binary_logloss: 0.569063
[168]	valid_0's auc: 0.764221	valid_0's binary_logloss: 0.569063
[169]	valid_0's auc: 0.764213	valid_0's binary_logloss: 0.56907
[170]	valid_0's auc: 0.764187	valid_0's binary_logloss: 0.569074
[171]	valid_0's auc: 0.764206	valid_0's binary_logloss: 0.569064
[172]	valid_0's auc: 0.764237	valid_0's binary_logloss: 0.569056
[173]	valid_0's auc: 0.764229	valid_0's binary_logloss: 0.569063
[174]	valid_0's auc: 0.76423

[292]	valid_0's auc: 0.764338	valid_0's binary_logloss: 0.569094
[293]	valid_0's auc: 0.764329	valid_0's binary_logloss: 0.569101
[294]	valid_0's auc: 0.764323	valid_0's binary_logloss: 0.5691
[295]	valid_0's auc: 0.76432	valid_0's binary_logloss: 0.569101
[296]	valid_0's auc: 0.764313	valid_0's binary_logloss: 0.569104
[297]	valid_0's auc: 0.764321	valid_0's binary_logloss: 0.569101
[298]	valid_0's auc: 0.764318	valid_0's binary_logloss: 0.569103
[299]	valid_0's auc: 0.76433	valid_0's binary_logloss: 0.569096
[300]	valid_0's auc: 0.764325	valid_0's binary_logloss: 0.569098
[301]	valid_0's auc: 0.764313	valid_0's binary_logloss: 0.5691
[302]	valid_0's auc: 0.764327	valid_0's binary_logloss: 0.569096
[303]	valid_0's auc: 0.76432	valid_0's binary_logloss: 0.569096
[304]	valid_0's auc: 0.764324	valid_0's binary_logloss: 0.569098


LGBMClassifier(bagging_fraction=0.5, bagging_freq=1, boosting='rf',
               is_training_metric=True, learning_rate=0.01,
               metric='binary_logloss', n_estimators=1000, objective='binary',
               seed=2020)

In [18]:
preds=model.predict(x_test)

In [19]:
pred_proba=model.predict_proba(test)

In [20]:
pred_proba.shape

(11383, 2)

In [21]:
submission= pd.read_csv('./sample_submission.csv')
submission['voted'] = pred_proba[:,1]
submission.to_csv('submission_proba.csv', index = False)

In [22]:
submission.head()

index     voted
0      0  0.669575
1      1  0.817942
2      2  0.496636
3      3  0.270775
4      4  0.751068

# F1 score

In [23]:
print('F1 score :{:.4f}'.format(f1_score(y_test, preds)))

F1 score :0.7065


# 파라미터 별 결과

### public / private / F1 : 0.779141352 / 0.770271404 / 0.6980 (최고점 160등)
- n_estimators=400

### public / private  F1 : 0.7773849453 / 0.7699346405 / 0.6959
- n_estimators=1000,
- learning_rate=0.01,
- max_depth=16,
- boosting='gbdt',
- objective='binary',
- metric='mse',
- is_training_metric=True,
- num_leaves=144,
- feature_fraction=0.9,
- bagging_fraction=0.7,
- bagging_freq=5,
- seed=2020

### public / private / F1 : 0.7783070588 / 0.7706610222 / 0.7013
- n_estimators=1000,
- learning_rate=0.01,
- boosting='gbdt',
- objective= 'binary',
- metric= 'auc',
- is_training_metric= True,
- seed=2020

### public / private / F1 : 0.7783070588 / 0.7706610222 / 0.7013
- n_estimators=1000,
- learning_rate=0.01,
- boosting='gbdt',
- objective= 'binary',
- metric= 'binary_logloss',
- is_training_metric= True,
- seed=2020

### public / private / F1 : 0.7770950256 / 0.7685265013 / 0.7043
- n_estimators=1000,
- learning_rate=0.01,
- boosting='dart',
- objective= 'binary',
- metric= 'binary_logloss',
- is_training_metric= True,
- seed=2020

### public / private / F1 : 0.77277671 / 0.7640015648 / 0.7065
- n_estimators=1000,
- learning_rate=0.01,
- boosting='rf',
- objective= 'binary',
- metric= 'binary_logloss',
- bagging_freq =1,
- bagging_fraction = 0.5,
- is_training_metric= True,
- seed=2020